ethergo/submitter/submitter.go

```diff
@@ -346,119 +346,211 @@ func (t *txSubmitterImpl) txTypeForChain(chainID *big.Int) (txType uint8) {
+// and gasPrice is copied to both gasFeeCap and gasTipCap in the opposite scenario.
+//
+//nolint:nestif
+func (t *txSubmitterImpl) populateGasFromPrevTx(ctx context.Context, transactor *bind.TransactOpts, prevTx *types.Transaction, currentDynamic bool) {
+	if prevTx == nil {
+		return
+	}
+
+	_, span := t.metrics.Tracer().Start(ctx, "submitter.populateGasFromPrevTx")
+
+	defer func() {
+		span.SetAttributes(
+			attribute.String("gas_price", bigPtrToString(transactor.GasPrice)),
+			attribute.String("gas_fee_cap", bigPtrToString(transactor.GasFeeCap)),
+			attribute.String("gas_tip_cap", bigPtrToString(transactor.GasTipCap)),
+		)
+		metrics.EndSpan(span)
+	}()
+
+	prevDynamic := prevTx.Type() == types.DynamicFeeTxType
+	if currentDynamic {
+		if prevDynamic {
+			transactor.GasFeeCap = core.CopyBigInt(prevTx.GasFeeCap())
+			transactor.GasTipCap = core.CopyBigInt(prevTx.GasTipCap())
+		} else {
+			transactor.GasFeeCap = core.CopyBigInt(prevTx.GasPrice())
+			transactor.GasTipCap = core.CopyBigInt(prevTx.GasPrice())
 		}
-		if transactor.GasFeeCap != nil && maxPrice.Cmp(transactor.GasFeeCap) < 0 {
-			transactor.GasFeeCap = maxPrice
+	} else {
+		if prevDynamic {
+			transactor.GasPrice = core.CopyBigInt(prevTx.GasFeeCap())
+		} else {
+			transactor.GasPrice = core.CopyBigInt(prevTx.GasPrice())
 		}
+	}
+}
 
+// applyGasFloor applies the min gas price from the config if values are unset.
+// Otherwise, gas values are bumped by the configured GasBumpPercentage.
+//
+//nolint:cyclop,nestif
+func (t *txSubmitterImpl) applyGasFloor(ctx context.Context, transactor *bind.TransactOpts, chainID int, useDynamic, shouldBump bool) {
+	_, span := t.metrics.Tracer().Start(ctx, "submitter.applyGasFloor")
+
+	defer func() {
 		span.SetAttributes(
 			attribute.String("gas_price", bigPtrToString(transactor.GasPrice)),
 			attribute.String("gas_fee_cap", bigPtrToString(transactor.GasFeeCap)),
 			attribute.String("gas_tip_cap", bigPtrToString(transactor.GasTipCap)),
+			attribute.Bool("should_bump", shouldBump),
 		)
-		metrics.EndSpanWithErr(span, err)
+		metrics.EndSpan(span)
 	}()
 
-	// TODO: cache both of these values
-	shouldBump := true
-	useEIP1559 := t.config.SupportsEIP1559(chainID)
-	if useEIP1559 {
-		transactor.GasFeeCap, err = client.SuggestGasPrice(ctx)
-		if err != nil {
-			return fmt.Errorf("could not get gas price: %w", err)
+	gasFloor := t.config.GetMinGasPrice(chainID)
+	if shouldBump {
+		bumpPct := t.config.GetGasBumpPercentage(chainID)
+		if useDynamic {
+			if transactor.GasFeeCap == nil {
+				transactor.GasFeeCap = gasFloor
+				span.AddEvent("gas fee cap is nil; setting to gas floor")
+			} else {
+				transactor.GasFeeCap = gas.BumpByPercent(transactor.GasFeeCap, bumpPct)
+			}
+			if transactor.GasTipCap == nil {
+				transactor.GasTipCap = gasFloor
+				span.AddEvent("gas tip cap is nil; setting to gas floor")
+			} else {
+				transactor.GasTipCap = gas.BumpByPercent(transactor.GasTipCap, bumpPct)
```

Comment:
Do we need to check `transactor.GasTipCap` for nil here?

Reply:
We check `if transactor.GasTipCap == nil {` just above :)

Reply:
Sorry, that was supposed to be a comment for this line:
https://github.com/synapsecns/sanguine/blob/6c52b8b698982134f37797897eb9fb8e63577264/ethergo/submitter/submitter.go#L459-L461